include/led.h

```diff
@@ -4,20 +4,109 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#ifndef _ZEPHYR_LED_STRIP_H
-#define _ZEPHYR_LED_STRIP_H
+#ifndef _ZEPHYR_LED_H
+#define _ZEPHYR_LED_H
 
 /**
  * @file
- * @brief Public API for controlling linear strips of LEDs.
- *
- * This library abstracts the chipset drivers for individually
- * addressable strips of LEDs.
+ * @brief Public API for controlling LEDs
  */
 
 #include <zephyr/types.h>
 #include <device.h>
 
+/**
+ * @typedef led_api_update_duty
+ * @brief Callback API for updating an LED duty cycle
+ *
+ * @see led_update_duty() for argument descriptions.
+ */
+typedef int (*led_api_update_duty)(struct device *dev, int id, int duty);
+
+/**
+ * @typedef led_api_update_freq
+ * @brief Callback API for updating an LED frequency
+ *
+ * @see led_update_freq() for argument descriptions.
+ */
+typedef int (*led_api_update_freq)(struct device *dev, int id, int freq);
+
+/**
+ * @typedef led_api_update_channel
+ * @brief Callback API for updating an LED channel
+ *
+ * @see led_update_channel() for argument descriptions.
+ */
+typedef int (*led_api_update_channel)(struct device *dev, int channel, int mode);
+
+/**
+ * @brief LED driver API
+ *
+ * This is the mandatory API that any LED driver needs to expose.
+ */
+struct led_driver_api {
+	led_api_update_duty update_duty;
+	led_api_update_freq update_freq;
+	led_api_update_channel update_channel;
+};
+
+/**
+ * @brief Update an LED device duty cycle
+ *
+ * This routine updates the LED device duty cycle
+ *
+ * @param dev LED device
+ * @param id ID of the LED duty cycle register
+ * @param duty Duty cycle value
+ * @return 0 on success, negative on error
+ */
+static inline int led_update_duty(struct device *dev,
+				    int id,
+				    int duty)
+{
+	const struct led_driver_api *api = dev->driver_api;
+
+	return api->update_duty(dev, id, duty);
+}
+
+/**
+ * @brief Update an LED device frequency
+ *
+ * This routine updates the LED device frequency
```

Comment:
This also could use some improvements. Is this frequency in Hertz, for example, or some other unit? What, in general, do you mean by the "frequency" of the LED device? (What is the periodic signal whose frequency is being set by this API?)

Why is it a plain signed int, instead of an unsigned type? I've only ever seen negative "frequencies" in the frequency domain representations of signals, but I don't think that's what you're referring to here.

Reply:
Agree. It should be unsigned and the units is cycles per second. Will make it more clear.